src/backend/cdb/cdbmutate.c

```diff
@@ -617,7 +617,8 @@ apply_motion(PlannerInfo *root, Plan *plan, Query *query)
 			{
 				if ((plan->flow->flotype == FLOW_PARTITIONED ||
 					(plan->flow->flotype == FLOW_SINGLETON &&
-					 plan->flow->locustype == CdbLocusType_SegmentGeneral)) &&
+					 plan->flow->locustype == CdbLocusType_SegmentGeneral) ||
+					 query->hasModifyingCTE) &&
```

Comment:
Initially, this condition said: if data on segments (partitioned or replicated flow) and we want it on coordinator (non-parallel cursor case), append gather motion. Why existing of Modifying CTE somewhere in plan leads to requirement to gather data unconditionally?

Reply:
Quick test without this condition.
[regression.diffs.txt](https://github.com/arenadata/gpdb/files/11755218/regression.diffs.txt)

The example query.
```
explain (costs off)
with cte as (
    update with_dml_dr set j = j + 1 where i <= 5 returning j
) select count(*) from cte;
             QUERY PLAN              
-------------------------------------
 Aggregate
   ->  Update on with_dml_dr
         ->  Seq Scan on with_dml_dr
               Filter: (i <= 5)
 Optimizer: Postgres query optimizer
(5 rows)

with cte as (
    update with_dml_dr set j = j + 1 where i <= 5 returning j
) select count(*) from cte;
 count 
-------
(0 rows)
```
For now, I can't remember all the reasons of why I decided to add this check and made it such wide.

Reply:
The `apply_motion` function is called in order to apply motion node to the top-level plan tree if necessary. For child plan nodes this function is not called.
As for modifying CTE's cases, `query->hasModifyingCTE` flag is set only for top-level select statement. And in the patch the `bringResultToDispatcher = true` will take place always when such CTE is present. However there is the finit set of possible Flow types at all: `FLOW_SINGLETON`, `FLOW_REPLICATED`, `FLOW_PARTITIONED`. 
For `FLOW_PARTITIONED` the condition will be true and it's ok independently from CTE presence.
For `FLOW_SINGLETON` there is the following logic:
1. **With our patch**:
`bringResultToDispatcher = true` because of `query->hasModifyingCTE` flag and `focusPlan` is [called](https://github.com/arenadata/gpdb/blob/6ed03a1d78ef8ab6560dc9a5b40a47cd75faaa75/src/backend/cdb/cdbmutate.c#L652-L680). However [inside](https://github.com/arenadata/gpdb/blob/6ed03a1d78ef8ab6560dc9a5b40a47cd75faaa75/src/backend/cdb/cdbllize.c#L1116-L1118) nothing changes. Then we [get](https://github.com/arenadata/gpdb/blob/6ed03a1d78ef8ab6560dc9a5b40a47cd75faaa75/src/backend/cdb/cdbmutate.c#L682)  to `apply_motion_mutator` function, where for most queries with modifying CTE (for examples in tests section of commit) this part of code is [executed](https://github.com/arenadata/gpdb/blob/6ed03a1d78ef8ab6560dc9a5b40a47cd75faaa75/src/backend/cdb/cdbmutate.c#L964-L969). Therefore, checking `query->hasModifyingCTE` in if-clause is really **unneseccary**  as it does not affect anything in this case.
2. **With our patch** we get to `apply_motion_mutator` without `focusPlan`, and have the same [situation](https://github.com/arenadata/gpdb/blob/6ed03a1d78ef8ab6560dc9a5b40a47cd75faaa75/src/backend/cdb/cdbmutate.c#L964-L969).
And `FLOW_REPLICATED` can happen in top-level plan node only for our modifying CTE case (unproven hypothesis), and we here we should add motion plan if `query->hasModifyingCTE` is true.

Therefore, `query->hasModifyingCTE` doesn't really break general behavior. And to make it a little bit clearer I can change the if-clause the following way:
```
if ((plan->flow->flotype == FLOW_PARTITIONED ||
   (plan->flow->flotype == FLOW_SINGLETON &&
   plan->flow->locustype == CdbLocusType_SegmentGeneral)) ||
   (plan->flow->flotype == FLOW_REPLICATED &&    <------ add check here
   query->hasModifyingCTE) && 
    !root->glob->is_parallel_cursor)
     bringResultToDispatcher = true;

```
I already ran the CI tests for version with such if-clause, tests were passing  without problems.

Reply:
I assume, that main questions are: Is the gather motion always explicit when the sender's locus type is Replicated? Will it be safe to remove such kind of restrictions on replicated locus? 
Well, consider the executor's code. [Here](https://github.com/arenadata/gpdb/blob/6ed03a1d78ef8ab6560dc9a5b40a47cd75faaa75/src/backend/executor/nodeMotion.c#L1056-L1067) it is said, that gather motion is marked as explicit for the locus Replicated. The concept of Explicit Gather Motion was added in this [commit](https://github.com/arenadata/gpdb/commit/fb7247b93bc43c534ae2adbaedb45d8947adefb2),which extended the abilities of locus Replicated, and this was done for RETURNING clause support in modifying queries. I guess, that all restrictions about relicated locus had been made before adding Explicit Gather Motion functionality, and the restrictions were weakened only for DML over replicated tables. And, imho, the other cases, where we need explicit data gathering, simply do not exists or irrelevant. (except modifying CTEs, of course).  I think, we can allow explicit gathering everywhere it is needed, because it's now safe.  Therefore, we can just add FLOW_RAPLICATED here without consequences.